### ISSUE TITLE
Not possible to use a local proxy

### ISSUE THREAD

**Issue by fipro78** [containers/podman-desktop#4549]
### Bug description

I am trying to install and setup Podman Desktop on my Windows machine at work. As it is my workplace machine, I am behind a firewall, and in the homeoffice connected via VPN. I have found the related documentations:

- https://podman-desktop.io/docs/proxy/using-a-proxy
- https://podman-desktop.io/docs/proxy/using-a-vpn-on-windows

For the VPN issue it took me a while to understand that the option __User mode networking (traffic relayed by a user process)__ is only available in the version for "restricted environments". Got that information only by surprise. Maybe it can be added to the VPN documentation that you need that version of the Podman Desktop installer to make it more clear.

But the proxy configuration is a bigger issue. If I configure the proxy like I am used to e.g. for Maven, Gradle or Docker Desktop with the protocol, I get the following error if I try to pull an image:

```
U:\>podman pull bash
Trying to pull docker.io/library/bash:latest...
Error: initializing source docker://bash:latest: pinging container registry registry-1.docker.io: Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp: lookup https: no such host
```

I suppose the idea of leaving out the protocol is for convenience, but it diverges from other proxy setting I am aware of. I knwo that some tools use the protocol, some don't. But at least having the protocol as part of the proxy setting never caused a failure. 

And it looks like, it avoids the usage of a __local proxy__ like [Px](https://github.com/genotrance/px). The reason is that even for HTTPS the proxy needs to be configured with `http://localhost:3128` which is not working with Podman Desktop.

Setting the proxy to `localhost:3128` or `127.0.0.1:3128` things are not working. One issue is of course the incorrect protocol, not sure if there is also the issue that `localhost` or `127.0.0.1` can't be translated from Podman inside the WSL to the local proxy running on the host. I get the following error in that case:

```
U:\>podman pull bash
Trying to pull docker.io/library/bash:latest...
Error: initializing source docker://bash:latest: pinging container registry registry-1.docker.io: Get "https://registry-1.docker.io/v2/": proxyconnect tcp: dial tcp 127.0.0.1:3128: connect: connection refused
```

With Docker Desktop the following configuration using a local proxy is working fine:

__Web Server (HTTP)__  
http://localhost:3128

__Secure Web Server (HTTPS)__  
http://localhost:3128

__Bypass proxy settings for these hosts & domains__  
host.docker.internal,localhost,127.0.0.1 

In Podman Desktop I have to apply the proxy in this format:

`<user>:<pwd>@<proxy>:<url>`

Especially setting the username and password in the URL not nice here and actually the reason for using a local proxy like Px, as this reduces the effort when the password needs to be changed.

And since today I also the the following error:

```
U:\>podman pull bash
Resolving "bash" using unqualified-search registries (/etc/containers/registries.conf.d/999-podman-machine.conf)
Trying to pull docker.io/library/bash:latest...
Error: initializing source docker://bash:latest: pinging container registry registry-1.docker.io: Get "https://registry-1.docker.io/v2/": Proxy Authentication Required
```

Although I added the credentials to the proxy URL and it worked last week this way. Also a `curl` from inside the __podman_machine_default__ works. So not sure why the pull suddently stopped working with the authentication error.

### Operating system

Windows 10

### Installation Method

Installer from website/GitHub releases

### Version

1.4.0

**Comment by n1hility** [comment]
@fipro78 hey you mentioned the curl works, does it work if you do a podman machine ssh, and from that prompt do a podman pull there? The way this works is the environment settings HTTP_PROXY/HTTPS_PROXY get replicated and should be set in the Fedora guest. When you use the curl command, where you using the proxy options on that, or where you seeing it work picking up the env?

**Comment by jeffmaury** [comment]
See #4356

**Comment by fipro78** [comment]
@n1hility 
Forget about the curl stuff and the authentication issue. It might be related to some local networking issue. I switched between the office (no VPN) and homeoffice (VPN). Interestingly the VPN works, but I didn't get it to work with the authentication this morning from the office. I will check it further.

@jeffmaury 
I downloaded the Windows installer from here https://github.com/containers/podman-desktop/actions/runs/6667995650?pr=4356

Whether setting the proxy to System or Manual doesn't have any effect. It only uses the values I enter in the fields. And the values need to be in format

    <user>:<pwd>@<url>:<port>

I also wonder how the System setting should work on Windows in my scenario. I am not sure if the Px proxy is actually a "system proxy" that is configured in the registry. I extracted it and configured it for autostart. Then I can use it via http://localhost:3128 in any other application. I even have the environment variables set on Windows, as they are also used by Windows applications (at least some).

If you have any idea or an update I could test, please let me know. With the version I installed now from the link I got sent, neither the System proxy setting nor a localhost or 127.0.0.1 is working.

But at least the prefix "http://" does not cause an error with the new version. So that seems to be fixed.

**Comment by jeffmaury** [comment]
I don't think PX is integrated with Windows proxy settings. This PR accept both format for proxy settings: either a full URL (thus you can insert the user and password if required) or the old style hostname:port.

Please note that PX has an issue with WSL that may explain you issue: https://github.com/genotrance/px/issues/173

**Comment by fipro78** [comment]
Not sure if this is really the issue. I think that the first issue is that localhost or respectively 127.0.0.1 is not resolved to the host, but is resolved as the WSL2 machine.  

I just opened this ticket because Docker Desktop on Windows is able to resolve the local proxy without any issues (now). But of course I don't know the technical details how they manage this. But Docker Desktop also has WSL2 distros running, therefore I guess it could be similar.

But maybe the answer then is simply "using a local proxy on the Windows host is not possible at the moment". And maybe it will be possible in the future once the `dnsTunneling=true` and `networkingMode=mirrored` settings are available in Windows. Unfortunately they are currently only available in the version from the insider program, which I don't have.

**Comment by fipro78** [comment]
@jeffmaury 
I notice an issue with the proxy settings. Our corporate proxy actualy requests the authentication via NTLM. If I pass the proxy url with username and password, every action (curl, dnf, podman) result in a "proxy authentication required" error. So the basic authentication seems to be not supported. I tested curl in the podman-machine WSL using this format

    curl --proxy-ntlm --proxy-user $Username:$Password --proxy $PROXY_URL:$PORT $SITE_URL

And the connect succeeded.

Then I configured /etc/dnf/dnf.conf and included the following settings:

    proxy=<url>:<port>
    proxy_username=<user>
    proxy_password=<pwd>
    proxy_auth_method=ntlm

and I was able to do a `sudo dnf update`

I then installed CNTLM into the podman machine and configured it accordingly, and after starting it I am able to pull images etc.

Not sure if this helps, but maybe it would be an option to install a local proxy to the podman machine if configured and edit the cntlm.conf based on the settings provided in the proxy settings of Podman Desktop.

BTW: I also noticed that sometimes the _/containers/containers.conf_ file gets corrupted after updating the proxy settings. Not sure if this is a known issue or something new for the 1.6.0-next version I am currently testing.

**Comment by fipro78** [comment]
Now that I got it working to pull images and start containers with the workaround above, I get to the next issue that the containers themselves can't connect to the internet via the proxy. Even setting the configuration according to this [page](https://podman-desktop.io/docs/proxy/using-a-proxy-in-your-containers) fails.

Interestingly the settings from the Podman Desktop seem to override the settings in the podman machine. And of course the container can not connect to the CNTLM instance running in the machine. At least I don't know which IP address I should use or what configuration might help.

**Comment by win5do** [comment]
I think I caught the bug.

```
macOS v13
podman-desktop version: v1.5.2
podman version: v4.7.2
```

Set up a proxy with the value "http://127.0.0.1:8080/"

![image](https://github.com/containers/podman-desktop/assets/22112659/4b525872-a3ad-47e0-a4c3-ee4d5a2d5b4e)

After applying the configuration and restarting the VM, enter the VM and check the env.
```sh
podman machine ssh

env | grep -i proxy

ouput:
http_proxy=http://host.containers.internal:8080
https_proxy=http://host.containers.internal:8080
HTTP_PROXY=http://http://host.containers.internal:8080
HTTPS_PROXY=http://http://host.containers.internal:8080
```

Upon closer inspection, the value of HTTP_PROXY has an extra "http://" prefix, while http_proxy does not.

The Go HTTP client encountered an incorrect proxy value, resulting in the error message "proxyconnect tcp: dial tcp: lookup http on 192.168.127.1:53: no such host."

Curl is able to work properly with the proxy, might be it using the correct value for "http_proxy".

**Comment by n1hility** [comment]
@fipro78

> But maybe the answer then is simply "using a local proxy on the Windows host is not possible at the moment". And maybe it will be possible in the future once the `dnsTunneling=true` and `networkingMode=mirrored` settings are available in Windows. Unfortunately they are currently only available in the version from the insider program, which I don't have.

Right mirrored isnt yet in the stable WSL stream, and requires some work on our end to support it. Although I very much think this is what we will ultimately use/require once it lands. It should resolve a number of these long-standing networking challenges with WSL. 

If you are running user-mode-networking, you can reach the windows host with a fixed ip address of "192.168.127.254".

With regular networking (and also user-mode-networking) you should also be able to user your PC name as a .local address (e.g. foo.local). Although your local system firewall policy would have to allow it.

Alternatively there are some hack solutions like a powershell script that updates a hostname entry.

In short, I think your workaround below of installing cntlm is better.

> @jeffmaury I notice an issue with the proxy settings. Our corporate proxy actualy requests the authentication via NTLM. 

 Ah yes. Does your corporate proxy support https urls? It might be just disallowing basic auth over http, in which case changing HTTP_PROXY and HTTPS_PROXY to use the encrypted version might help (basic over TLS is safe). Although they may have just denied it across the board. 

The reason this doesn't work out of the box, is that the proxy support in podman comes from Golang's http client implementation, and golang does not have NTLM support.  I created an issue to investigate adding this to podman in the future:

 - https://github.com/containers/podman/issues/20656
 
> I then installed CNTLM into the podman machine and configured it accordingly, and after starting it I am able to pull images etc.

This is a good idea for a workaround. 

> Not sure if this helps, but maybe it would be an option to install a local proxy to the podman machine if configured and edit the cntlm.conf based on the settings provided in the proxy settings of Podman Desktop.

We can consider this as well as an alternative to #20656

> Now that I got it working to pull images and start containers with the workaround above, I get to the next issue that the containers themselves can't connect to the internet via the proxy. Even setting the configuration according to this [page](https://podman-desktop.io/docs/proxy/using-a-proxy-in-your-containers) fails.
> 
> Interestingly the settings from the Podman Desktop seem to override the settings in the podman machine. And of course the container can not connect to the CNTLM instance running in the machine. At least I don't know which IP address I should use or what configuration might help.

You should be able to get to it using `host.containers.internal` as the hostname to reach the linux host running cntlm. Does that work for you?

**Comment by n1hility** [comment]
> ```shell
> http_proxy=http://host.containers.internal:8080
> https_proxy=http://host.containers.internal:8080
> HTTP_PROXY=http://http://host.containers.internal:8080
> HTTPS_PROXY=http://http://host.containers.internal:8080
> ```
> 
> Upon closer inspection, the value of HTTP_PROXY has an extra "http://" prefix, while http_proxy does not.

@win5do Good find! Created a new issue for this other problem: https://github.com/containers/podman-desktop/issues/4765

**Comment by fipro78** [comment]
> You should be able to get to it using `host.containers.internal` as the hostname to reach the linux host running cntlm. Does that work for you?

I added the following to /etc/containers/containers.conf as the machine is a rootful machine:

```
[containers]
http_proxy = true
env = ["http_proxy=host.containers.internal:3128", "https_proxy=host.containers.internal:3128"]
```

But that setting doesn't seem to have an effect. It gets overriden by the proxy settings from Podman Desktop.

**Comment by n1hility** [comment]
> > You should be able to get to it using `host.containers.internal` as the hostname to reach the linux host running cntlm. Does that work for you?
> 
> I added the following to /etc/containers/containers.conf as the machine is a rootful machine:
> 
> ```
> [containers]
> http_proxy = true
> env = ["http_proxy=host.containers.internal:3128", "https_proxy=host.containers.internal:3128"]
> ```
> 
> But that setting doesn't seem to have an effect. It gets overriden by the proxy settings from Podman Desktop.

@fipro78 
Thanks for trying. This may be counter-intuitive, but try setting` http_proxy = false`, but keep your env settings. This will cause it to ignore the env variables that podman desktop is setting, using instead the ones you are manually setting.

**Comment by corburn** [comment]
In the past, podman-desktop setup worked for me, but now fails with:
```
Error: Get "https://builds.coreos.fedoraproject.org/streams/testing.json": proxyconnect tcp: dial tcp: lookup https: no such host
```

The code used to assume the `httpsProxy` did not include the `https://` scheme prefix. I wonder if the change introduced by PR #4825 is why `HTTPS_PROXY` has the duplicated scheme prefix.

**Comment by corburn** [comment]
I think I found it: https://github.com/containers/podman-desktop/blob/73958b53e8a3d573b5bb63efe4aee0371bb3c384/packages/main/src/plugin/util/exec.ts#L56-L59
